### PR TITLE
API GW CLI Updates

### PR DESCRIPTION
**Pull request by mdeuser**
- Initial CLI changes to support API GW
- Negative usage tests for now (positive tests awaiting local API GW - coming soon)

### CODE REVIEW

**Comment by mdeuser** [comment]
@dubeejw - Do you have time for a cursory code review?  Especially consider the error response processing updates.

**Comment by mdeuser** [comment]
PG1 861 - CLI tests passed

**Comment by mdeuser** [comment]
Last Travis failure due to unrelated test case
```
system.basic.WskBasicTests > Wsk Action CLI should create and invoke a blocking action resulting in an error response object FAILED
    org.scalatest.exceptions.TestFailedException: "error: Unable to invoke action 'errorResponseObject': The connection failed, or timed out. (HTTP status code 502)
    " did not include substring that matched regex "error": "name '!' contains illegal characters \(code \d+\)"
```

File: tools/cli/go-whisk-cli/wski18n/resources/en_US.all.json [hunk 1]
```diff
@@ -1150,5 +1150,169 @@
...
+  {
+    "id":"'{{.name}}' is not a valid action name: {{.err}}",
+    "translation": "'{{.name}}' is not a valid action name: {{.err}}"
+  },
+  {
+    "id":"'{{.name}}' is not a valid action name.",
+    "translation": "'{{.name}}' is not a valid action name."
+  },
+  {
+    "id": "Unable to create api: {{.err}}",
```
Comment (dubeejw) [inline tools/cli/go-whisk-cli/wski18n/resources/en_US.all.json]: Capitalize API.

File: tools/cli/go-whisk-cli/wski18n/resources/en_US.all.json [hunk 2]
```diff
@@ -1150,5 +1150,169 @@
...
+    "translation": "Unable to create api: {{.err}}"
```
Comment (dubeejw) [inline tools/cli/go-whisk-cli/wski18n/resources/en_US.all.json]: Capitalize API.

File: tools/cli/go-whisk-cli/wski18n/resources/en_US.all.json [hunk 3]
```diff
@@ -1150,5 +1150,169 @@
...
+  },
+  {
+    "id": "{{.ok}} created api {{.path}} {{.verb}} for action {{.name}}\n{{.fullpath}}\n",
```
Comment (dubeejw) [inline tools/cli/go-whisk-cli/wski18n/resources/en_US.all.json]: Capitalize API.

File: tools/cli/go-whisk-cli/wski18n/resources/en_US.all.json [hunk 4]
```diff
@@ -1150,5 +1150,169 @@
...
+    "translation": "{{.ok}} created api {{.path}} {{.verb}} for action {{.name}}\n{{.fullpath}}\n"
```
Comment (dubeejw) [inline tools/cli/go-whisk-cli/wski18n/resources/en_US.all.json]: Capitalize API.

File: tools/cli/go-whisk-cli/wski18n/resources/en_US.all.json [hunk 5]
```diff
@@ -1150,5 +1150,169 @@
...
+  },
+  {
+    "id": "Unable to parse api command arguments: {{.err}}",
```
Comment (dubeejw) [inline tools/cli/go-whisk-cli/wski18n/resources/en_US.all.json]: Capitalize API.

File: tools/cli/go-whisk-cli/wski18n/resources/en_US.all.json [hunk 6]
```diff
@@ -1150,5 +1150,169 @@
...
+    "translation": "Unable to parse api command arguments: {{.err}}"
```
Comment (dubeejw) [inline tools/cli/go-whisk-cli/wski18n/resources/en_US.all.json]: Capitalize API.

File: tools/cli/go-whisk-cli/wski18n/resources/en_US.all.json [hunk 7]
```diff
@@ -1150,5 +1150,169 @@
...
+  {
+    "id": "create a new API",
+    "translation": "create a new API"
+  },
+  {
+    "id": "update an existing API",
+    "translation": "update an existing API"
+  },
+  {
+    "id": "Unable to update api: {{.err}}",
```
Comment (dubeejw) [inline tools/cli/go-whisk-cli/wski18n/resources/en_US.all.json]: Capitalize API.

File: tools/cli/go-whisk-cli/wski18n/resources/en_US.all.json [hunk 8]
```diff
@@ -1150,5 +1150,169 @@
...
+    "translation": "Unable to update api: {{.err}}"
```
Comment (dubeejw) [inline tools/cli/go-whisk-cli/wski18n/resources/en_US.all.json]: Capitalize API.

File: tools/cli/go-whisk-cli/wski18n/resources/en_US.all.json [hunk 9]
```diff
@@ -1150,5 +1150,169 @@
...
+  },
+  {
+    "id": "{{.ok}} updated api {{.path}} {{.verb}} for action {{.name}}\n{{.fullpath}}\n",
```
Comment (dubeejw) [inline tools/cli/go-whisk-cli/wski18n/resources/en_US.all.json]: Capitalize API.

File: tools/cli/go-whisk-cli/wski18n/resources/en_US.all.json [hunk 10]
```diff
@@ -1150,5 +1150,169 @@
...
+    "translation": "{{.ok}} updated api {{.path}} {{.verb}} for action {{.name}}\n{{.fullpath}}\n"
```
Comment (dubeejw) [inline tools/cli/go-whisk-cli/wski18n/resources/en_US.all.json]: Capitalize API.

File: tools/cli/go-whisk-cli/wski18n/resources/en_US.all.json [hunk 11]
```diff
@@ -1150,5 +1150,169 @@
...
+  },
+  {
+    "id": "get API",
+    "translation": "get API"
+  },
+  {
+    "id": "Unable to get api: {{.err}}",
```
Comment (dubeejw) [inline tools/cli/go-whisk-cli/wski18n/resources/en_US.all.json]: Capitalize API.

File: tools/cli/go-whisk-cli/wski18n/resources/en_US.all.json [hunk 12]
```diff
@@ -1150,5 +1150,169 @@
...
+    "translation": "Unable to get api: {{.err}}"
```
Comment (dubeejw) [inline tools/cli/go-whisk-cli/wski18n/resources/en_US.all.json]: Capitalize API.

File: tools/cli/go-whisk-cli/wski18n/resources/en_US.all.json [hunk 13]
```diff
@@ -1150,5 +1150,169 @@
...
+  },
+  {
+    "id": "{{.ok}} api for path {{.path}} verb {{.verb}}\n",
```
Comment (dubeejw) [inline tools/cli/go-whisk-cli/wski18n/resources/en_US.all.json]: Capitalize API.

File: tools/cli/go-whisk-cli/wski18n/resources/en_US.all.json [hunk 14]
```diff
@@ -1150,5 +1150,169 @@
...
+    "translation": "{{.ok}} api for path {{.path}} verb {{.verb}}\n"
```
Comment (dubeejw) [inline tools/cli/go-whisk-cli/wski18n/resources/en_US.all.json]: Capitalize API.

File: tools/cli/go-whisk-cli/wski18n/resources/en_US.all.json [hunk 15]
```diff
@@ -1150,5 +1150,169 @@
...
+  {
+    "id": "delete an API",
+    "translation": "delete an API"
+  },
+  {
+    "id": "Unable to delete action: {{.err}}",
+    "translation": "Unable to delete action: {{.err}}"
+  },
+  {
+    "id": "{{.ok}} deleted api {{.basepath}}\n",
```
Comment (dubeejw) [inline tools/cli/go-whisk-cli/wski18n/resources/en_US.all.json]: Capitalize API.

File: tools/cli/go-whisk-cli/wski18n/resources/en_US.all.json [hunk 16]
```diff
@@ -1150,5 +1150,169 @@
...
+    "translation": "{{.ok}} deleted api {{.basepath}}\n"
```
Comment (dubeejw) [inline tools/cli/go-whisk-cli/wski18n/resources/en_US.all.json]: Capitalize API.

File: tools/cli/go-whisk-cli/wski18n/resources/en_US.all.json [hunk 17]
```diff
@@ -1150,5 +1150,169 @@
...
+  {
+    "id": "{{.ok}} deleted {{.path}} {{.verb}} from {{.basepath}}\n",
+    "translation": "{{.ok}} deleted {{.path}} {{.verb}} from {{.basepath}}\n"
+  },
+  {
+    "id": "list APIs",
+    "translation": "list APIs"
+  },
+  {
+    "id": "Unable to obtain the list of apis: {{.err}}",
```
Comment (dubeejw) [inline tools/cli/go-whisk-cli/wski18n/resources/en_US.all.json]: Capitalize API.

File: tools/cli/go-whisk-cli/wski18n/resources/en_US.all.json [hunk 18]
```diff
@@ -1150,5 +1150,169 @@
...
+    "translation": "Unable to obtain the list of apis: {{.err}}"
```
Comment (dubeejw) [inline tools/cli/go-whisk-cli/wski18n/resources/en_US.all.json]: Capitalize API.

File: tools/cli/go-whisk-cli/wski18n/resources/en_US.all.json [hunk 19]
```diff
@@ -1150,5 +1150,169 @@
...
+  {
+    "id": "API collection `NAME` (default NAMESPACE)",
+    "translation": "API collection `NAME` (default NAMESPACE)"
+  },
+  {
+    "id": "The API `BASE_PATH` to which the API_PATH is relative",
+    "translation": "The API `BASE_PATH` to which the API_PATH is relative"
+  },
+  {
+    "id": "{{.ok}} apis\n",
```
Comment (dubeejw) [inline tools/cli/go-whisk-cli/wski18n/resources/en_US.all.json]: Capitalize API.

File: tools/cli/go-whisk-cli/wski18n/resources/en_US.all.json [hunk 20]
```diff
@@ -1150,5 +1150,169 @@
...
+    "translation": "{{.ok}} apis\n"
```
Comment (dubeejw) [inline tools/cli/go-whisk-cli/wski18n/resources/en_US.all.json]: Capitalize API.

File: tools/cli/go-whisk-cli/wski18n/resources/en_US.all.json [hunk 21]
```diff
@@ -1150,5 +1150,169 @@
...
+  {
+    "id": "A configuration file was not specified.",
+    "translation": "A configuration file was not specified."
+  },
+  {
+    "id": "Error parsing swagger file '{{.name}}': {{.err}}",
+    "translation": "Error parsing swagger file '{{.name}}': {{.err}}"
+  },
+  {
+    "id": "Swagger file is invalid (missing basePath, info, paths, or swagger fields",
```
Comment (dubeejw) [inline tools/cli/go-whisk-cli/wski18n/resources/en_US.all.json]: Close parenthesis.

File: tools/cli/go-whisk-cli/wski18n/resources/en_US.all.json [hunk 22]
```diff
@@ -1150,5 +1150,169 @@
...
+    "translation": "Swagger file is invalid (missing basePath, info, paths, or swagger fields"
```
Comment (dubeejw) [inline tools/cli/go-whisk-cli/wski18n/resources/en_US.all.json]: Close parenthesis.

File: tools/cli/go-whisk-cli/wski18n/resources/en_US.all.json [hunk 23]
```diff
@@ -1150,5 +1150,169 @@
...
+  {
+    "id": "list APIs",
+    "translation": "list APIs"
+  },
+  {
+    "id": "Unable to obtain the list of apis: {{.err}}",
+    "translation": "Unable to obtain the list of apis: {{.err}}"
+  },
+  {
+    "id": "'{{.verb}}' is not a valid API verb.  Valid values are: {{.verbs}}",
```
Comment (dubeejw) [inline tools/cli/go-whisk-cli/wski18n/resources/en_US.all.json]: Single space after period.

File: tools/cli/go-whisk-cli/wski18n/resources/en_US.all.json [hunk 24]
```diff
@@ -1150,5 +1150,169 @@
...
+    "translation": "'{{.verb}}' is not a valid API verb.  Valid values are: {{.verbs}}"
```
Comment (dubeejw) [inline tools/cli/go-whisk-cli/wski18n/resources/en_US.all.json]: Single space after period.

File: tools/cli/go-whisk-cli/wski18n/resources/en_US.all.json [hunk 25]
```diff
@@ -1150,5 +1150,169 @@
...
+  {
+    "id": "{{.ok}} apis\n",
+    "translation": "{{.ok}} apis\n"
+  },
+  {
+    "id": "{{.url}} {{.operation}} {{.action}}\n",
+    "translation": "{{.url}} {{.operation}} {{.action}}\n"
+  },
+  {
+    "id": "An API base path is required.  An optional API relative path and operation may also be provided.",
```
Comment (dubeejw) [inline tools/cli/go-whisk-cli/wski18n/resources/en_US.all.json]: Single space after period.

File: tools/cli/go-whisk-cli/wski18n/resources/en_US.all.json [hunk 26]
```diff
@@ -1150,5 +1150,169 @@
...
+    "translation": "An API base path is required.  An optional API relative path and operation may also be provided."
```
Comment (dubeejw) [inline tools/cli/go-whisk-cli/wski18n/resources/en_US.all.json]: Single space after period.

File: tools/cli/go-whisk/whisk/api.go [hunk 27]
```diff
@@ -0,0 +1,272 @@
+/*
+ * Copyright 2015-2016 IBM Corporation
```
Comment (dubeejw) [inline tools/cli/go-whisk/whisk/api.go]: 2016

File: tools/cli/go-whisk/whisk/api.go [hunk 28]
```diff
@@ -0,0 +1,272 @@
...
+    Activated       bool      `json:"gwApiActivated"`
+    TenantId        string    `json:"tenantId"`
+    Swagger         *ApiSwagger `json:"apidoc,omitempty"`
+}
+
+type ApiSwagger struct {
+    SwaggerName     string    `json:"swagger,omitempty"`
+    BasePath        string    `json:"basePath,omitempty"`
+    Info            *ApiSwaggerInfo `json:"info,omitempty"`
+    Paths           map[string]map[string]map[string]map[string]interface{} `json:"paths,omitempty"`
```
Comment (dubeejw) [inline tools/cli/go-whisk/whisk/api.go]: Whoa

File: tools/cli/go-whisk/whisk/api.go [hunk 29]
```diff
@@ -0,0 +1,272 @@
...
+    "OPTIONS": true,
+}
+
+////////////////////
+// Api Methods //
+////////////////////
+
+func (s *ApiService) List(apiListOptions *ApiListOptions) (*RetApiReponseApiArray, *http.Response, error) {
+    var route string
+    route = fmt.Sprintf("routes")
```
Comment (dubeejw) [inline tools/cli/go-whisk/whisk/api.go]: route = "routes"

Comment (dubeejw) [inline tools/cli/go-whisk/whisk/api.go]: Or just pass "routes" into addRouteOptions.

File: tools/cli/go-whisk/whisk/api.go [hunk 30]
```diff
@@ -0,0 +1,272 @@
...
+        return nil, resp, err
+    }
+
+    return apiArray, resp, err
+}
+
+func (s *ApiService) Insert(api *SendApi, overwrite bool) (*RetApiReponse, *http.Response, error) {
+    var sentAction interface{}
+
+    route := fmt.Sprintf("routes")
```
Comment (dubeejw) [inline tools/cli/go-whisk/whisk/api.go]: fmt.Sprintf() not needed.

**Comment by dubeejw** [comment]
@mdeuser, I left some comments on things that would be easy to fix before you merge. I will still need more time to trace through the API command code.

File: tools/cli/go-whisk/whisk/api.go (hunk 28 above)
Comment (mdeuser) [inline tools/cli/go-whisk/whisk/api.go]: @dubeejw - hopefully there's a more readable way to dig into a nest JSON without having to define structures for every intermediate object in the nested path.

**Comment by mdeuser** [comment]
@dubeejw - Thanks for the review thus far.  The updated PR includes updates for the current comments

**Comment by csantanapr** [comment]
Thanks @dubeejw  and @mdeuser I will merge this part of the APIGW, full function will not be there but that's the plan to do it in stages.

**Comment by csantanapr** [comment]
@mdeuser lets do a PG

**Comment by mdeuser** [comment]
PG2 681 successful

**Pull request merged**